File: solutions/elixir/change/1/lib/change.ex
defmodule Change do
  @doc """
    Determine the least number of coins to be given to the user such
    that the sum of the coins' value would equal the correct amount of change.
    It returns {:error, "cannot change"} if it is not possible to compute the
    right amount of coins. Otherwise returns the tuple {:ok, list_of_coins}

    ## Examples

      iex> Change.generate([5, 10, 15], 3)
      {:error, "cannot change"}

      iex> Change.generate([1, 5, 10], 18)
      {:ok, [1, 1, 1, 5, 10]}

  """
  @error {:error, "cannot change"}

  @spec generate(list, integer) :: {:ok, list} | {:error, String.t()}
  def generate(_coins, target)
      when target === 0,
      do: {:ok, []}

  def generate([smallest_coin | _more_coins], target)
      when target < smallest_coin,
      do: @error

  def generate(coins, target),
    do:
      coins
      |> Enum.reverse()
      |> generate(target, _accumulator = [])

  def generate(coins, target, _accumulator)
      when length(coins) === 0 and
             target !== 0,
      do: @error

  def generate(coins, _target, accumulator)
      when length(coins) === 0,
      do: {:ok, accumulator}

  def generate([largest_coin | more_coins], target, accumulator)
      when largest_coin > target,
      do: generate(more_coins, target, accumulator)

  def generate([largest_coin | _more_coins] = coins, target, accumulator) do
    obvious_result =
      {obvious_status, obvious_collection} =
      generate(coins, target - largest_coin, [largest_coin | accumulator])

    if obvious_status === :ok && length(obvious_collection) === 1,
      do: {:ok, obvious_collection},
      else: generate_alternatives(coins, target, accumulator, obvious_result)
  end

  def generate_alternatives([_largest_coin | more_coins], target, accumulator, {:error, _}),
    do: generate(more_coins, target, accumulator)

  def generate_alternatives(
        [largest_coin | more_coins],
        target,
        accumulator,
        {:ok, obvious_collection}
      ) do
    {alternative_status, alternative_collection} =
      more_coins
      |> Enum.drop_while(&(rem(largest_coin, &1) === 0))
      |> generate(target, accumulator)

    if alternative_status === :error ||
         length(alternative_collection) >= length(obvious_collection),
       do: {:ok, obvious_collection},
       else: {:ok, alternative_collection}
  end
end

File: solutions/erlang/change/1/src/change.erl
-module(change).

-export([find_fewest_coins/2]).

-define(ERROR, {error, "cannot change"}).

find_fewest_coins(Target, Coins)
  when Target < 0
  -> erlang:error(Target, Coins);

find_fewest_coins(Target, _Coins)
  when Target == 0
  -> [];

find_fewest_coins(Target, [SmallestCoin | _MoreCoins])
  when Target < SmallestCoin
  -> undefined;

find_fewest_coins(Target, Coins)
  ->
  case find_fewest_coins(Target, lists:reverse(Coins), _Accumulator = []) of
    {ok, FewestCoins} -> FewestCoins;
    {error, _} -> undefined
  end.

find_fewest_coins(Target, Coins, _Accumulator)
  when (Target =/= 0) andalso (length(Coins)) == 0
  -> ?ERROR;

find_fewest_coins(_Target, Coins, Accumulator)
  when length(Coins) == 0
  -> {ok, Accumulator};

find_fewest_coins(Target, [LargestCoin | MoreCoins], Accumulator)
  when LargestCoin > Target
  -> find_fewest_coins(Target, MoreCoins, Accumulator);

find_fewest_coins(Target, [LargestCoin | _MoreCoins] = Coins, Accumulator) ->
  ObviousResult = {ObviousStatus, ObviousCollection} =
    find_fewest_coins(Target - LargestCoin, Coins, [LargestCoin | Accumulator]),

  if
    (ObviousStatus == ok)
      and (length(ObviousCollection) == 1) ->
      {ok, ObviousCollection};

    true ->
      find_fewest_coins_alternatives(Target, Coins, Accumulator, ObviousResult)
  end.

find_fewest_coins_alternatives(Target, [_LargestCoin | MoreCoins], Accumulator, {error, _})
  -> find_fewest_coins(Target, MoreCoins, Accumulator);

find_fewest_coins_alternatives(Target, [LargestCoin | MoreCoins], Accumulator, {ok, ObviousCollection})
  ->
  AvailableCoins = lists:dropwhile(fun(Coin) -> LargestCoin rem Coin == 0 end, MoreCoins),
  {AlternativeStatus, AlternativeCollection} = find_fewest_coins(Target, AvailableCoins, Accumulator),

  NoAlternative = AlternativeStatus == error,
  AlternativeTooBig = length(AlternativeCollection) >= length(ObviousCollection),
  if
    NoAlternative or AlternativeTooBig ->
      {ok, ObviousCollection};

    true ->
      {ok, AlternativeCollection}
  end.
